semeio/jobs/correlated_observations_scaling/job.py

```diff
@@ -1,232 +1,104 @@
-    data points per node.
-    """
-    for event in events:
-        obs_vector = obs[event.key]
-        for index, obs_node in enumerate(obs_vector):
-            if obs_vector.getImplementationType().name == "SUMMARY_OBS":
-                index_list = (
-                    event.index if event.index is not None else range(len(obs_vector))
-                )
-                if index in index_list:
-                    obs_node.set_std_scaling(scale_factor)
-            elif obs_vector.getImplementationType().name != "SUMMARY_OBS":
-                obs_node.updateStdScaling(scale_factor, event.active_list)
-    print(
-        "Keys: {} scaled with scaling factor: {}".format(
-            [event.key for event in events], scale_factor
+from semeio.jobs.correlated_observations_scaling.scaled_matrix import DataMatrix
+from semeio.jobs.correlated_observations_scaling.validator import has_keys, is_subset
+
+
+class ScalingJob(object):
+    def __init__(self, obs_keys, obs, obs_with_data, user_config_dict):
+        """Creates a ScalingJob instance with the given obs_keys, obs,
+        obs_with_data and user_config_dict."""
+        self._obs = obs
+        self._obs_with_data = obs_with_data
+        self._obs_keys = obs_keys
+        self._validation_errors = None
+        self._config = self._setup_configuration(user_config_dict)
+
+    def scale(self, measured_data):
+        """
+        Collects data, performs scaling and applies scaling, assumes validated input.
+        """
+        config = self._config.snapshot
+
+        measured_data.remove_failed_realizations()
+        measured_data.remove_inactive_observations()
+        measured_data.filter_ensemble_mean_obs(config.CALCULATE_KEYS.alpha)
+        measured_data.filter_ensemble_std(config.CALCULATE_KEYS.std_cutoff)
+
+        matrix = DataMatrix(measured_data.data)
+        matrix.std_normalization(inplace=True)
+
+        scale_factor = matrix.get_scaling_factor(config.CALCULATE_KEYS)
+
+        update_data = create_active_lists(self._obs, config.UPDATE_KEYS.keys)
+        self._update_scaling(self._obs, scale_factor, update_data)
+
+    def valid(self):
+        """
+        Returns whether or not this job is valid. If invalid, the set of error
+        messages can be retrieved using get_errors().
+        """
+        errors = [] if self._config.valid else self._config.errors
+        calc_keys = self.get_calc_keys()
+        application_keys = [
+            entry.key for entry in self._config.snapshot.UPDATE_KEYS.keys
+        ]
+
+        errors.extend(is_subset(calc_keys, application_keys))
+        errors.extend(
+            has_keys(self._obs_keys, calc_keys, "Key: {} has no observations")
+        )
+        errors.extend(has_keys(self._obs_with_data, calc_keys, "Key: {} has no data"))
+        self._validation_errors = errors
+        return len(self._validation_errors) == 0
+
+    def get_errors(self):
+        if self._validation_errors is None:
```

Comment:
This seems slightly off to me. Why should I call `valid` before `get_errors`. I would suggest the following; we implement to properties `valid` and `errors` and ensure validation before these are called. We can either do this at creation time (I don't really see why we don't want to do that) and compute `_errors` then. Or we can lazily compute the errors the first time either `valid` or `errors` is called. But the later discussion is an implementation detail really...

Reply:
Yeah no real reason. Will fix.

Reply:
This is my bad, it already _did_ raise an exception in the _initializer_ (just a function). However, I refactored this because I didn't want the `ScalingJob` to decide what to do with the errors. It used to be printed right in the `valid_job` function, but it needs to be propagated to the top. There it can be printed or whatever.

I will re-introduce the exception, but allow the `errors` to be carried on that exception (so a subclass of `ValueError` called `ValidationError`. OK?

Reply:
I don't think the two are exclusive are they? I do however agree to leave it to the caller to deal with the errors occurring. Whether you introduce an exception or error code convention for propagating failures I leave to you 👍